File: benchmarks-QuTiP/expect_operator.py
import qutip as qt
import numpy as np
import benchmarkutils

name = "expect_operator"

samples = 5
evals = 100
cutoffs = range(100, 2501, 100)

def setup(N):
    op = (qt.destroy(N) + qt.create(N))
    psi = qt.Qobj(np.ones(N, complex)/(N**(1/2)))
    rho = psi*psi.dag()
    return op, rho

def f(op, rho):
    return qt.expect(op, rho)

print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
checks = {}
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    op, rho = setup(N)
    checks[N] = f(op, rho)
    t = benchmarkutils.run_benchmark(f, op, rho, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)

File: benchmarks-QuantumOptics.jl/expect_operator.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

name = "expect_operator"

samples = 5
evals = 100
cutoffs = [100:100:2500;]

function f(op, state)
    expect(op, state)
end

println("Benchmarking: ", name)
print("Cutoff: ")
checks = Dict{Int, Float64}()
results = []
for N in cutoffs
    print(N, " ")
    b = FockBasis(N-1)
    op = (destroy(b) + create(b))
    psi = Ket(b, ones(Complex128, N)/sqrt(N))
    rho = psi ⊗ dagger(psi)
    checks[N] = abs(f(op, rho))
    t = @belapsed f($op, $rho) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)
